### ISSUE TITLE
Create an S3 bucket for storing assessment images

### ISSUE THREAD

**Issue by jsf9k** [cisagov/cool-system#176]
# 💡 Summary #

Create an S3 bucket that COOL assessors can use for storing and retrieving their assessment VM images.

## Motivation and context ##

A few different assessors have asked for this over the past year.  Apparently downloading the images from Ripley is an exercise in patience, so the need is approaching criticality.  :radioactive: 

## Implementation notes ##

@dav3r and @jsf9k discussed this briefly and sketched out the following solution:
1. An S3 VPC gateway endpoint is added to the Shared Services VPC (where the COOL VPN lives).  _Because of the way the VPC routing is configured, all S3 traffic from the VPN users will egress via this S3 VPC gateway endpoint._  This code lives in the [cisagov/cool-sharedservices-networking](https://github.com/cisagov/cool-sharedservices-networking) repository.  (Note that this is one piece of #197.)
1. Modify [the OpenVPN configuration](https://github.com/cisagov/openvpn-server-tf-module/blob/develop/cloudinit/openvpn-config.tpl.yml#L45-L51) to push a route to clients so that all S3 traffic is routed through the VPN.  The destination CIDRs for S3 traffic can be determined via `aws ec2 describe-prefix-lists --filters Name=prefix-list-name,Values=com.amazonaws.us-east-1.s3`, for example, and those CIDRS can be added to the `private_networks` input variable of [cisagov/openvpn-server-tf-module](https://github.com/cisagov/openvpn-server-tf-module) when it is called via [cisagov/cool-sharedservices-openvpn](https://github.com/cisagov/cool-sharedservices-openvpn).
1. S3 buckets named `cisa-cool-assessment-images-production` and `cisa-cool-assessment-images-staging` are created in the production and staging `Images` AWS accounts, respectively.  This code lives in a new repository named [cisagov/cool-images-assessment-images](https://github.com/cisagov/cool-images-assessment-images).
1. The `cisa-cool-assessment-images-production` and `cisa-cool-assessment-images-staging` buckets are configured with [a policy that _only_ allows access via the S3 VPC gateway endpoint in the Shared Services account](https://docs.aws.amazon.com/AmazonS3/latest/userguide/example-bucket-policies-vpc-endpoint.html#example-bucket-policies-restrict-accesss-vpc-endpoint).  This code lives in [cisagov/cool-images-assessment-images](https://github.com/cisagov/cool-images-assessment-images) and uses the remote state of [cisagov/cool-sharedservices-networking](https://github.com/cisagov/cool-sharedservices-networking).
1. The `cisa-cool-assessment-images-production` and `cisa-cool-assessment-images-staging` buckets are configured with no other access restrictions.  If you can access the bucket via the Shared Services S3 VPC gateway endpoint (i.e. if you are connected to the COOL VPN) then you can access the bucket contents.  

## Tasks ##

- [x] Create new repo `cisagov/cool-images-assessment-images` to contain Terraform below
- [x] Write and apply Terraform in `cisagov/cool-images-assessment-images` to:
  - [x] Create `cisa-cool-assessment-images-staging` bucket in `Images (Staging)` account
  - [x] Create `cisa-cool-assessment-images-production` bucket in `Images (Production)` account
  - [x] Create `AssessmentImagesBucketFullAccess` policy and role in each `Images` account
 allowing read/write/delete access to buckets above
  - [x] Create `assessment-images-bucket-full-access` user in `Users` account
  - [x] Create `Images-AssumeAssessmentImagesBucketFullAccess` policy in `Users` account allowing assumption of `AssessmentImagesBucketFullAccess` role in `Images (Staging)` and `Images (Production)` and attach it to `assessment-images-bucket-full-access` user
  - [x] Create bucket policy to allow read-only access by COOL VPN users (via AWS CLI)
  - [x] Add necessary S3 routes to the OpenVPN instance via [cisagov/cool-sharedservices-openvpn](https://github.com/cisagov/cool-sharedservices-openvpn)
- [ ] Manually delete the artisanal AWS resources created during proof-of-concept testing:
  - [ ] `Images (Staging)` account:
    - [ ] `policy/TEMP-ImagesTestingWrite`
    - [ ] `role/TEMP-ImagesTestingBucketWrite`
  - [ ] `Users` account:
    - [ ] `policy/TEMP-Images-AssumeImagesTestingBucketWrite`
    - [ ] `user/temp-images-bucket-write`

Nice to have, but not required for initial MVP:
  - [ ] ~~Make buckets keep only `n` versions of each object~~ This is not supported by AWS at this time. Only time-based aging of objects is currently supported as mentioned in https://github.com/cisagov/cool-system/issues/176#issuecomment-823523705

## Acceptance criteria ##

- [ ] Implementation described above is, uh, implemented
- [ ] Ability to access bucket contents by COOL VPN users is verified
- [ ] Inability to access bucket contents via any other means is verified

**Comment by felddy** [comment]
This is related to VM import as well.  See #181

**Comment by jsf9k** [comment]
@felddy says that it would be better if this data didn't flow over the VPN.

**Comment by bra1ncramp** [comment]
We definitely need to E1 proof this. We want to be able to pre-configure systems to allow the user to grab this without their manually configuring each system. Please note that pen testers will be given a different laptop for each engagement, and all data from the previous user MUST be removed in-between engagements. 
Pre-signed URLS are ok, but we would need a way to securely, consistently, predictably, and conveniently transmit URLs.

**Comment by mcdonnnj** [comment]
Just to comment on version management:
It appears that currently S3 buckets only support version expiration based on the age of the non-current version per
https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-lifecycle-mgmt.html
and with Terraform
https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/s3_bucket#noncurrent_version_expiration

The `noncurrent_version_expiration` object supports the following
 * `days` (Required) Specifies the number of days noncurrent object versions expire.

**Comment by bra1ncramp** [comment]
I'm not sure this is completed. We still have our setup using the testing S3 folder.

**Comment by mcdonnnj** [comment]
@bra1ncramp Thanks for noticing. I accidentally linked instead of just mentioning the issue so it closed out when I merged in the initial implementation. We will need to have the setup switched to test the newly created resources before we can tear down the testing resources and close this out.

**Comment by dav3r** [comment]
@mcdonnnj and/or @bra1ncramp - can you please provide a status update on this effort?

**Comment by bra1ncramp** [comment]
The VMs were pushed from out end to both Production and Staging. 
@mcdonnnj , could you confirm that you received the VMs?

**Comment by mcdonnnj** [comment]
The VMs are indeed uploaded to the buckets. I am working toward implementing the functionality that was deemed MVP-necessary on Friday (the previous issue wording was poor) to allow anyone on the VPN read access to the contents of these buckets.

**Comment by mcdonnnj** [comment]
The bucket policies have been implemented and I verified functionality in the staging environment. However, this functionality in the production environment will not be available until the OpenVPN instance for the environment is redeployed.

**Comment by mcdonnnj** [comment]
With OpenVPN instances redeployed and functionality confirmed this should be considered fully implemented.
